### PR TITLE
Data hash - only run cycle on relevant changes

### PR DESCRIPTION
**Pull request by alewitt2**

### CODE REVIEW

File: lib/BaseController.js [hunk 1]
```diff
@@ -194,21 +266,48 @@ module.exports = class BaseController {
...
+    objectPath.set(patchObj, ['razee-logs', logLevel, logHash], log);
+    this._razeeLogHashes.push(logHash);
+
+    let res = await this.patchSelf({ status: patchObj }, { status: true });
+    // save newly patched object to continue cycle with latest data
+    objectPath.set(this._data, 'object', res);
+    return res;
+  }
+
+  async _patchRazeeLogs() { // clear out logs in status that weren't created this cycle
```
Comment (adamkingit) [inline lib/BaseController.js]: Rename method

File: lib/BaseController.js [hunk 2]
```diff
@@ -108,42 +108,107 @@ module.exports = class BaseController {
...
+    let lockClusterPath = './config/lock-cluster';
+    let exists = await fs.pathExists(lockClusterPath);
+    if (exists) {
+      lockCluster = await fs.readFile(lockClusterPath, 'utf8');
+      lockCluster = lockCluster.trim().toLowerCase();
+    }
+    return (lockCluster == 'true');
+  }
+
+  computeDataHash(resource) {
```
Comment (adamkingit) [inline lib/BaseController.js]: Refactor to make it simpler for subclasses to identify `importantData`

File: lib/BaseController.js [hunk 3]
```diff
@@ -108,42 +108,107 @@ module.exports = class BaseController {
   // the handler calls the underscored event function to allow pre/post processesing
   // around the normal event function that should be overriden in the subclass
   async _added() {
+    // should always keep data-hash up to date on added events to avoid conflict with modified events
+    this._logger.debug(`Updating annotation deploy.razee.io/data-hash for ADDED event.. ${objectPath.get(this._data, 'object.metadata.resourceVersion')}`);
+    let res = await this.patchSelf({ metadata: { annotations: { 'deploy.razee.io/data-hash': this.computeDataHash(objectPath.get(this._data, 'object')) } } });
```
Comment (adamkingit) [inline lib/BaseController.js]: Check hash before patching

File: lib/CompositeController.js [hunk 4]
```diff
@@ -44,17 +51,33 @@ module.exports = class CompositeController extends BaseController {
     return res;
   }
 
-  addChildren(children) {
+  async addChildren(children) {
```
Comment (adamkingit) [inline lib/CompositeController.js]: eliminate duplication

File: lib/BaseController.js (hunk 3 above)
Comment (alewitt2) [reply lib/BaseController.js]: done

File: lib/BaseController.js (hunk 2 above)
Comment (alewitt2) [reply lib/BaseController.js]: done

File: lib/CompositeController.js (hunk 4 above)
Comment (alewitt2) [reply lib/CompositeController.js]: done

**Pull request merged**